metrics-core/src/main/java/com/codahale/metrics/LockFreeExponentiallyDecayingReservoir.java

```diff
@@ -0,0 +1,244 @@
+
+    private LockFreeExponentiallyDecayingReservoir(int size, double alpha, Duration rescaleThreshold, Clock clock) {
+        // Scale alpha to nanoseconds
+        this.alphaNanos = alpha * SECONDS_PER_NANO;
+        this.size = size;
+        this.clock = clock;
+        this.rescaleThresholdNanos = rescaleThreshold.toNanos();
+        this.state = new State(clock.getTick(), new AtomicLong(), new ConcurrentSkipListMap<>());
+    }
+
+    @Override
+    public int size() {
+        return (int) Math.min(size, state.count.get());
+    }
+
+    @Override
+    public void update(long value) {
+        long now = clock.getTick();
+        rescaleIfNeeded(now).update(value, now);
+    }
+
+    private State rescaleIfNeeded(long currentTick) {
+        State stateSnapshot = this.state;
+        long lastScaleTick = stateSnapshot.startTick;
+        if (currentTick - lastScaleTick >= rescaleThresholdNanos) {
+            State newState = stateSnapshot.rescale(currentTick);
+            if (stateUpdater.compareAndSet(this, stateSnapshot, newState)) {
+                // newState successfully installed
+                return newState;
+            }
+            // Otherwise another thread has won the race and we can return the result of a volatile read.
+            // It's possible this has taken so long that another update is required, however that's unlikely
+            // and no worse than the standard race between a rescale and update.
+            return this.state;
+        }
+        return stateSnapshot;
+    }
+
+    @Override
+    public Snapshot getSnapshot() {
+        State stateSnapshot = rescaleIfNeeded(clock.getTick());
+        return new WeightedSnapshot(stateSnapshot.values.values());
+    }
+
+    private double weight(long durationNanos) {
+        return Math.exp(alphaNanos * durationNanos);
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * By default this uses a size of 1028 elements, which offers a 99.9%
+     * confidence level with a 5% margin of error assuming a normal distribution, and an alpha
+     * factor of 0.015, which heavily biases the reservoir to the past 5 minutes of measurements.
+     */
+    public static final class Builder {
+        private static final int DEFAULT_SIZE = 1028;
+        private static final double DEFAULT_ALPHA = 0.015D;
+        private static final Duration DEFAULT_RESCALE_THRESHOLD = Duration.ofHours(1);
+
+        private int size = DEFAULT_SIZE;
+        private double alpha = DEFAULT_ALPHA;
+        private Duration rescaleThreshold = DEFAULT_RESCALE_THRESHOLD;
+        private Clock clock = Clock.defaultClock();
+
+        private Builder() {}
+
+        /**
+         * Maximum number of samples to keep in the reservoir. Once this number is reached older samples are
+         * replaced (based on weight, with some amount of random jitter).
+         */
+        public Builder size(int value) {
+            this.size = value;
```

Comment:
may want to validate size, not sure it makes sense to allow empty reservoirs?

```suggestion
            Preconditions.checkArgument(value > 0, "size must be greater than zero");
            this.size = value;
```

Reply:
addressed by https://github.com/dropwizard/metrics/pull/1656/commits/9518a8bc4cbe31929a0ef5d836e85307a7e7a590

Reply:
Good call, updated https://github.com/dropwizard/metrics/pull/1656/commits/9518a8bc4cbe31929a0ef5d836e85307a7e7a590